File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    if (!head || !head.next) return head;
    let dummy = new ListNode()
    let current = head // 1
    let previous = dummy // D
    
    while (current && current.next) {
        previous.next = current.next // D -> 2
        current.next = previous.next.next // 1 -> 3
        previous.next.next = current // 2 -> 1
        
        
        previous = current // previous = 1
        current = current.next // current = 3
    }
    
    return dummy.next;
};

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        dummyHead = ListNode()
        current = head
        previous = dummyHead
        
        while current and current.next:
            previous.next = current.next
            current.next = current.next.next
            previous.next.next = current
            
            previous = current
            current = current.next
            
        return dummyHead.next